### PR TITLE
For flyweights, shift from llvm::SmallVector to a type that enforces index types.

### PR DESCRIPTION
**Pull request by jonmeow**
The intent here is to reduce use of vanilla `int32_t` without a clear indicator of what it's referencing, and to more tightly link references with the underlying types they reference into.

The type name DataIndex doesn't feel great, but I was kind of floundering for a better name.

### CODE REVIEW

**Comment by chandlerc** [comment]
We were supposed to chat about this with @zygoloid today but I forgot -- let's try to remember to do so as I remain somewhat torn between the two fundamental directions possible with this PR -- fully typechecked modeling vs. lightweight modeling with a "decorated" integer.

**Comment by chandlerc** [comment]
Chatted with @zygoloid about this, and he leaned towards creating a custom *container* being a step too far. Mostly I think this stems from composability concerns and the API growth that would follow from that.

But we're both happy with having a template utility to provide a really good index type that catches as much misuse as we can manage.

I'll start reviewing the actual code with that direction in mind, but let me know if you want to make any adjustments first.

**Comment by jonmeow** [comment]
WDYT of this approach, then?

File: toolchain/common/data_index.h [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    data.push_back(el);
+    return index;
+  }
+
+  // Returns what the index will be if an element is manually added to the data,
+  // after this call.
+  template <typename VectorT>
+  static auto BeforeManualAppend(const VectorT& data) -> DataIndex {
+    return DataIndex(data.size());
+  }
```
Comment (chandlerc) [inline toolchain/common/data_index.h]: @zygoloid actually mentioned exactly this kind of approach, along with the `In` methods below, but neither of us were very happy with it.

I think its just a bit too much of an idiomatic break from how vectors typically work.

The design I think he was suggesting was much more around keeping the implicit conversion to an integer type, and just trying to make sure that going from `DataIndex<T>` <-> `DataIndex<U>` is prevented.

**Review by chandlerc — CHANGES_REQUESTED** [review]

File: toolchain/common/data_index.h (hunk 1 above)
Comment (jonmeow) [reply toolchain/common/data_index.h]: Okay, but I don't understand why you described that in terms of catching misuse. That is, at best, just reducing keyboard typing (versus the already-merged approach).

So does this new approach work for you? If not, I think this is reaching bare minimum levels of utility, so I'm thinking about just closing the PR.

File: toolchain/common/index_base.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+// DataIndex is designed to be passed by value, not reference or pointer. They
+// are also designed to be small and efficient to store in data structures.
+struct IndexBase {
+  IndexBase() : index(-1) {}
+  explicit IndexBase(int index) : index(index) {}
+
+  auto Print(llvm::raw_ostream& output) const -> void { output << index; }
+
+  int32_t index;
+};
```
Comment (chandlerc) [inline toolchain/common/index_base.h:31]: Any particular reason to not make this convertible, that would seem more convenient to me?

I also personally would find it a bit less surprising when reading users of this to have this be private and use an accessor method (maybe in addition to an implicit conversion for use with `[]`s). Any reason you prefer this?

**Review by chandlerc — APPROVED** [review]
A couple of questions about this design below. FWIW, approving and happy to change these in follow-ups if useful. I'm even happy to write a follow-up as I have another minor change I'm thinking about just to tidy things up.

One question I have for my follow-up, is there a reason to not go ahead and make all of the relational operators available?

File: toolchain/common/index_base.h (hunk 2 above)
Comment (jonmeow) [reply toolchain/common/index_base.h:31]: Style seems to make a strong statement here, "Do not define implicit conversions." https://google.github.io/styleguide/cppguide.html#Implicit_Conversions

"Implicit conversions can sometimes be necessary and appropriate for types that are designed to be interchangeable, for example when objects of two types are just different representations of the same underlying value." -- but I'm not convinced that these should be treated as interchangeable. The cost of `.index` feels low to me.

e.g., I think the line math is awkward and shouldn't magically/transparently work. Although really, maybe the line value should be a number unlike everything else, representing how it's really the line number, not some abstract index.

Comment (jonmeow) [reply toolchain/common/index_base.h:31]: Probably notable too that I would still lean towards _more_ type safety, not less.

**Comment by jonmeow** [comment]
> One question I have for my follow-up, is there a reason to not go ahead and make all of the relational operators available?

Most code in the toolchain is probably going to view it as the former. Most code _does not_ need relational operators. So to turn this around, why add support for things that most code doesn't need?

From my perspective, it encourages treating these values as just integers, not abstract index values. I was uncomfortable with it before, but it kinda sorta made sense when I was getting more type safety, and I was still thinking about how to deal with semantics. At this point, it'd just be adding more integer-like functionality to what semantics wants to just be abstract values; that's where I start to see the breakdown of an abstraction, that the library might be counter-productive.

**Comment by chandlerc** [comment]
> > One question I have for my follow-up, is there a reason to not go ahead and make all of the relational operators available?
> 
> Most code in the toolchain is probably going to view it as the former. Most code _does not_ need relational operators. So to turn this around, why add support for things that most code doesn't need?

One, we have them in more than one place already. Two, I find it a more complete abstraction model. But that gets to the point below...

> 
> From my perspective, it encourages treating these values as just integers, not abstract index values. I was uncomfortable with it before, but it kinda sorta made sense when I was getting more type safety, and I was still thinking about how to deal with semantics. At this point, it'd just be adding more integer-like functionality to what semantics wants to just be abstract values; that's where I start to see the breakdown of an abstraction, that the library might be counter-productive.

I don't really follow this. My analogy is a random-access iterator. It isn't an integer at all. But you can use relational operators and talk about one being "less" or "before" the other.

I think indices, including fully type safe and abstract indices, should similarly support random access idioms. For me, this isn't about them being *integers*, this is about them being *indices* into a data structure. Now, those aren't completely separable, I understand that. The mental model for me of an "indexable data structure" (in C++, there are other domains with other meanings of "index") is intrinsically tied to the index forming an offset into the structure, which is somewhat inherently integer-like. But I'm not trying to make these *be* integers, I'm just trying to get the abstract semantic model of an index.

The implicit conversion thing is, IMO, totally separate and maybe its too much overhead. I certainly wouldn't want it to be used anywhere *other* than inside of `[]`s for all the reasons you mention about it being useful to not treat these as integers.

**Comment by jonmeow** [comment]
> > One question I have for my follow-up, is there a reason to not go ahead and make all of the relational operators available?
> 
> Most code in the toolchain is probably going to view it as the former. Most code _does not_ need relational operators. So to turn this around, why add support for things that most code doesn't need?
> 
> From my perspective, it encourages treating these values as just integers, not abstract index values. I was uncomfortable with it before, but it kinda sorta made sense when I was getting more type safety, and I was still thinking about how to deal with semantics. At this point, it'd just be adding more integer-like functionality to what semantics wants to just be abstract values; that's where I start to see the breakdown of an abstraction, that the library might be counter-productive.

I should add, it may be fair to add another base type that does relational operators for the few types that want it. But I do think it's the minority that have relational operators right now, and won't grow by much more even as more index types are added, just because relational operators aren't generally useful for the work being done.

**Pull request merged**